{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ec52e13",
   "metadata": {},
   "source": [
    "# PBMCs Tutorial\n",
    "PBMC stands for Peripheral Blood Mononuclear Cells, a vital group of immune system cells found in the blood"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b78e00a4-c9eb-4bc9-8d89-7821091201e9",
   "metadata": {},
   "source": [
    "## 0. Initial setup\n",
    "- Install LINGER according to the [GitHub](https://github.com/Durenlab/LINGER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e732c1fd-ad8b-42d7-ab5b-4bd0b7f7c03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# conda create -n LINGER python==3.10.0\n",
    "# conda activate LINGER\n",
    "# pip install LingerGRN==1.105\n",
    "# conda install -c bioconda bedtools  # Requirement"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "999eb39c-709e-4124-a714-c714b1427eb1",
   "metadata": {},
   "source": [
    "- Register the kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b286245-3063-4898-9fb0-8e69442f876b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# pip install ipykernel\n",
    "# python -m ipykernel install --user --name LINGER --display-name \"Python (LINGER)\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7484e5-189d-448d-a091-a6cb096d3e7c",
   "metadata": {},
   "source": [
    "- Verify installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f9a5b095-db61-4fd1-8d8f-9aff689da21e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: LingerGRN\n",
      "Version: 1.105\n",
      "Summary: Gene regulatory network inference\n",
      "Home-page: https://github.com/Durenlab/LINGER\n",
      "Author: Kaya Yuan\n",
      "Author-email: qyyuan33@gmail.com\n",
      "License: MIT\n",
      "Location: /home/users/v/a/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages\n",
      "Requires: anndata, joblib, matplotlib, numpy, pandas, pybedtools, rpy2, scanpy, scikit-learn, scipy, seaborn, shap, statsmodels, torch, umap-learn\n",
      "Required-by: \n"
     ]
    }
   ],
   "source": [
    "!pip show LingerGRN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8f2d61fd-1575-45ac-b922-8e288665f0ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# conda environments:\n",
      "#\n",
      "base                     /home/ucl/inma/vangysel/miniconda3\n",
      "LINGER                *  /home/ucl/inma/vangysel/miniconda3/envs/LINGER\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda info --envs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b013a13f-3d6b-4767-a4aa-0856bee711c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda list -n LINGER                    # lists all packages in LINGER env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "22c56842-6a72-4a74-856c-6d691023c7d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /home/ucl/inma/vangysel/miniconda3/envs/LINGER:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "lingergrn                 1.105                    pypi_0    pypi\n"
     ]
    }
   ],
   "source": [
    "!conda list -n LINGER LingerGRN          # look for package LingerGRN in LINGER env"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94cee155-8bbc-4781-8462-26d298fcecb2",
   "metadata": {},
   "source": [
    "- Check ressources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1476e02b-544b-46ab-8236-de5cadefd55e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "256\n"
     ]
    }
   ],
   "source": [
    "!nproc --all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "53495ce9-84dd-4759-81a9-3c94ce08163d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              total        used        free      shared  buff/cache   available\n",
      "Mem:          754Gi        73Gi       572Gi       2.9Gi       108Gi       673Gi\n",
      "Swap:         4.0Gi       1.3Gi       2.7Gi\n"
     ]
    }
   ],
   "source": [
    "!free -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1488580d-c7eb-4f5e-b2ea-28de4670d29a",
   "metadata": {},
   "source": [
    "## 1. Download the general gene regulatory network\n",
    "This is the pretrained NN on bulk multiomics data across tissues : RNA-seq (gene expr.) and ATAC-seq (chrom. acc.) that will then be fine tuned with our single cell data. There is one pretrained NN per gene.\n",
    "\n",
    "### About the bulk GRN\n",
    "\n",
    "It contains three types of interactions (TF-RE-TG) : \n",
    "- TF &rarr; RE : biding strength (&alpha;)\n",
    "- RE &rarr; TG : cis regulatory strength (&beta;)\n",
    "- TF &rarr; TG : trans regulatory strength (&gamma;)\n",
    "\n",
    "We obtain **&alpha;** by extracting the weights from the input layer to the second layer (each TF and RE are connected to the 64 hidden neurons of h1). An embedding of a TF/RE is a vector of weitghs, we can then measure how similar two embeddings are. If a TF and RE have similar learned representations, they are likely to interact and will have a high biding strength.<br><br>\n",
    "We get **&beta;** and **&gamma;** using the average shapley value (that calculates the contribution of a feature to the prediction) over all cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ea0e4a80-49c7-4c14-970b-e259179737f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/users/v/a/vangysel/linger\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dfa9da48-7a14-48f0-9d42-ff1c06d5bd79",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2026-02-05 13:11:05 URL:https://drive.usercontent.google.com/download?export=download&confirm=&id=1jwRgRHPJrKABOk7wImKONTtUupV7yJ9b [20812483490/20812483490] -> \"data_bulk.tar.gz\" [1]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Set directories and download general GRN\n",
    "Datadir=$GLOBALSCRATCH/LINGER_data                     \n",
    "mkdir -p $Datadir   \n",
    "    \n",
    "# Download general GRN from Google Drive\n",
    "wget -nv -O $Datadir --load-cookies /tmp/cookies.txt \"https://drive.usercontent.google.com/download?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://drive.usercontent.google.com/download?id=1jwRgRHPJrKABOk7wImKONTtUupV7yJ9b'  -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\\1\\n/p')&id=1jwRgRHPJrKABOk7wImKONTtUupV7yJ9b\" -O data_bulk.tar.gz\n",
    "rm -rf /tmp/cookies.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03a7cb13-1651-4ba6-b9df-ff0a47d059f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "!tar -xzf data_bulk.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc1052a8-1a16-490d-92dd-1e2ccc3ec366",
   "metadata": {},
   "source": [
    "## 2. Prepare the input data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca3a34d4-96ce-4bcf-bdb2-e722d8213ded",
   "metadata": {},
   "source": [
    "- Download the h5 file (the matrix contains both RNA and ATAC data combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0748bf09-c8f8-412c-8ef0-5539db181785",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2026-02-04 15:24:27--  https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5\n",
      "Resolving cf.10xgenomics.com (cf.10xgenomics.com)... 104.18.0.173, 104.18.1.173, 2606:4700::6812:1ad, ...\n",
      "Connecting to cf.10xgenomics.com (cf.10xgenomics.com)|104.18.0.173|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 162282142 (155M) [binary/octet-stream]\n",
      "Saving to: ‘data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5’\n",
      "\n",
      "data/pbmc_granulocy 100%[===================>] 154.76M   109MB/s    in 1.4s    \n",
      "\n",
      "2026-02-04 15:24:29 (109 MB/s) - ‘data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5�� saved [162282142/162282142]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir -p data\n",
    "wget --progress=bar:force -O data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5 https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e5c9c60-71f1-4615-9132-af39747a6801",
   "metadata": {},
   "source": [
    "- Download cell annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "88ae6b10-8f2f-4124-9fb2-9eb3a7c507c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2026-02-04 15:31:32--  https://docs.google.com/uc?export=download&confirm=&id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_\n",
      "Resolving docs.google.com (docs.google.com)... 74.125.206.138, 74.125.206.113, 74.125.206.102, ...\n",
      "Connecting to docs.google.com (docs.google.com)|74.125.206.138|:443... connected.\n",
      "HTTP request sent, awaiting response... 303 See Other\n",
      "Location: https://drive.usercontent.google.com/download?id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_&export=download [following]\n",
      "--2026-02-04 15:31:33--  https://drive.usercontent.google.com/download?id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_&export=download\n",
      "Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 64.233.166.132, 2a00:1450:400c:c09::84\n",
      "Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|64.233.166.132|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 362958 (354K) [application/octet-stream]\n",
      "Saving to: ‘PBMC_label.txt’\n",
      "\n",
      "PBMC_label.txt      100%[===================>] 354.45K  --.-KB/s    in 0.01s   \n",
      "\n",
      "2026-02-04 15:31:36 (27.1 MB/s) - ‘PBMC_label.txt’ saved [362958/362958]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "wget --progress=bar:force -O data --load-cookies /tmp/cookies.txt \"https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\\1\\n/p')&id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_\" -O PBMC_label.txt && rm -rf /tmp/cookies.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eb3c3900-c946-476f-8f7e-d7f860afef47",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/anndata/_core/anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11909 × 144978\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import pandas as pd\n",
    "\n",
    "adata = sc.read_10x_h5('data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5', gex_only=False)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2282412e-fc46-41f6-a4f9-ff6da8d3c73d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(adata.var.index) - len(adata.var.index.unique()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "696b9ce6-5ca8-449c-9fb5-2b9dca79d353",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names_make_unique()     # only 10 var are not unique, we make them unique [USEFULL ??]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "112a5721-6e64-45da-aed5-62f39a078084",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix = adata.X.T    # linger expects k_features x n_cells\n",
    "adata.var['gene_ids'] = adata.var.index\n",
    "\n",
    "# features are genes and peaks grouped together (col1 for gene/peak name and col2 for category: gene or peak)\n",
    "features = pd.DataFrame(adata.var['gene_ids'].values.tolist(),columns=[1])\n",
    "features[2] = adata.var['feature_types'].values\n",
    "\n",
    "barcodes = pd.DataFrame(adata.obs_names,columns=[0])\n",
    "label = pd.read_csv('data/PBMC_label.txt',sep='\\t',header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a3f0d965-1c58-43ab-865d-bffb95c13e1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/LingerGRN/preprocess.py:278: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_RNA.obs['label']=label.loc[adata_RNA.obs['barcode']]['label'].values\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/LingerGRN/preprocess.py:281: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_ATAC.obs['label']=label.loc[adata_ATAC.obs['barcode']]['label'].values\n"
     ]
    }
   ],
   "source": [
    "from LingerGRN.preprocess import *\n",
    "adata_RNA, adata_ATAC = get_adata(matrix,features,barcodes,label)     # adata_RNA and adata_ATAC are scRNA and scATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f4274b70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Features : \n",
      "\n",
      "             1                2\n",
      "0  MIR1302-2HG  Gene Expression\n",
      "1      FAM138A  Gene Expression\n",
      "2        OR4F5  Gene Expression\n",
      "3   AL627309.1  Gene Expression\n",
      "4   AL627309.3  Gene Expression\n",
      "                             1      2\n",
      "144973  KI270713.1:20444-22615  Peaks\n",
      "144974  KI270713.1:27118-28927  Peaks\n",
      "144975  KI270713.1:29485-30706  Peaks\n",
      "144976  KI270713.1:31511-32072  Peaks\n",
      "144977  KI270713.1:37129-37638  Peaks\n",
      "\n",
      "\n",
      "Barcodes: \n",
      "                    0\n",
      "0  AAACAGCCAAGGAATC-1\n",
      "1  AAACAGCCAATCCCTT-1\n",
      "2  AAACAGCCAATGCGCT-1\n",
      "3  AAACAGCCACACTAAT-1\n",
      "4  AAACAGCCACCAACCG-1\n",
      "\n",
      "\n",
      "Labels : \n",
      "                           barcode_use               label\n",
      "barcode_use                                               \n",
      "AAACAGCCAAGGAATC-1  AAACAGCCAAGGAATC-1   naive CD4 T cells\n",
      "AAACAGCCAATCCCTT-1  AAACAGCCAATCCCTT-1  memory CD4 T cells\n",
      "AAACAGCCAATGCGCT-1  AAACAGCCAATGCGCT-1   naive CD4 T cells\n",
      "AAACAGCCAGTAGGTG-1  AAACAGCCAGTAGGTG-1   naive CD4 T cells\n",
      "AAACAGCCAGTTTACG-1  AAACAGCCAGTTTACG-1  memory CD4 T cells\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"Features : \\n\\n{features.head()}\")\n",
    "print(f\"{features.tail()}\\n\\n\")\n",
    "print(f\"Barcodes: \\n{barcodes.head()}\\n\\n\")\n",
    "print(f\"Labels : \\n{label.head()}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efbb20f8",
   "metadata": {},
   "source": [
    "### 2.1 About the `get_data()` function\n",
    "**@inputs :** \n",
    "- matrix: sparse matrix with RNA and ATAC data stacked vertically\n",
    "- features: gene/peak IDs and their types ('Gene Expression' or 'Peaks')\n",
    "- barcodes: cell barcodes\n",
    "- label: cell type labels/annotations \n",
    "\n",
    "**@outputs :**\n",
    "- adata_RNA \n",
    "- adata_ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "38c8caad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_ids             MIR1302-2HG\n",
      "feature_types    Gene Expression\n",
      "genome                    GRCh38\n",
      "Name: MIR1302-2HG, dtype: object\n",
      "\n",
      "Series([], Name: AAACAGCCAAGGAATC-1, dtype: float64)\n"
     ]
    }
   ],
   "source": [
    "# gene - cell \n",
    "print(adata.var.iloc[0], end=\"\\n\\n\")\n",
    "print(adata.obs.iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4d0f79a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11909 × 144978\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "af033110",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAGGAATC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAATCCCTT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAATGCGCT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACACTAAT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCAACCG-1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACAGCCAAGGAATC-1, AAACAGCCAATCCCTT-1, AAACAGCCAATGCGCT-1, AAACAGCCACACTAAT-1, AAACAGCCACCAACCG-1]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5a8cb20f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ids    feature_types  genome\n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  GRCh38\n",
       "FAM138A          FAM138A  Gene Expression  GRCh38\n",
       "OR4F5              OR4F5  Gene Expression  GRCh38\n",
       "AL627309.1    AL627309.1  Gene Expression  GRCh38\n",
       "AL627309.3    AL627309.3  Gene Expression  GRCh38"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9965f21b-8387-4b35-8ffb-5ecf9b8b66f7",
   "metadata": {},
   "source": [
    "## 4. About the AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ca0c9730",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11909 × 144978\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata            # cells x (genes + peaks) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1a36c8b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 9543 × 36601\n",
       "    obs: 'barcode', 'sample', 'label', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt'\n",
       "    var: 'gene_ids', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA        # cells x genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8c103b4e-c479-4d09-824a-485f5d170213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 9543 × 108377\n",
       "    obs: 'barcode', 'sample', 'label'\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC       # cells x peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e031dd6d-510d-441a-b1b7-7683eb89d401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adata_RNA.X[i, j]             # cell i, gene j : Gene expression count of gene j in cell i\n",
    "#adata_ATAC.X[i, k]            # cell i, peak k : Chromatin accessibility count of peak k in cell i\n",
    "adata_RNA.X[1, 44]             # cell 1 gene 44"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e3aa6436",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=float32)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.X[0, :10].toarray()     # cell 0, expr. of 10 first genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "03656578",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]], dtype=float32)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.X[:10, 0].toarray()     # gene 0 expr. in 10 first cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1d96f463",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample</th>\n",
       "      <th>label</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>AGCACTAGTTACGCGG-1</td>\n",
       "      <td>1</td>\n",
       "      <td>memory CD4 T cells</td>\n",
       "      <td>1474</td>\n",
       "      <td>2765.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4670</th>\n",
       "      <td>CGGCCATAGGGACTAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>memory B cells</td>\n",
       "      <td>1980</td>\n",
       "      <td>4381.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>GACGCCTAGGGTGAGT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>CD56 (bright) NK cells</td>\n",
       "      <td>1537</td>\n",
       "      <td>2934.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4694</th>\n",
       "      <td>CGGGCTTAGAGGAGTC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>classical monocytes</td>\n",
       "      <td>1475</td>\n",
       "      <td>2783.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6716</th>\n",
       "      <td>GCATCCTTCAATCATG-1</td>\n",
       "      <td>1</td>\n",
       "      <td>CD56 (dim) NK cells</td>\n",
       "      <td>2520</td>\n",
       "      <td>6102.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>AAGTGCAAGCCAAATC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD4 T cells</td>\n",
       "      <td>1619</td>\n",
       "      <td>3398.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3087</th>\n",
       "      <td>CAATCCTGTCATAACG-1</td>\n",
       "      <td>1</td>\n",
       "      <td>CD56 (bright) NK cells</td>\n",
       "      <td>3081</td>\n",
       "      <td>7406.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5119</th>\n",
       "      <td>CTAGGACGTGTTCCCA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>plasmacytoid DC</td>\n",
       "      <td>3108</td>\n",
       "      <td>7025.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>ACGCCACAGTAGCTTA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>memory B cells</td>\n",
       "      <td>1907</td>\n",
       "      <td>3958.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7808</th>\n",
       "      <td>GGGTGTTGTTCGCTCA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>classical monocytes</td>\n",
       "      <td>1874</td>\n",
       "      <td>3686.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 barcode  sample                   label  n_genes_by_counts  \\\n",
       "1506  AGCACTAGTTACGCGG-1       1      memory CD4 T cells               1474   \n",
       "4670  CGGCCATAGGGACTAA-1       1          memory B cells               1980   \n",
       "6010  GACGCCTAGGGTGAGT-1       1  CD56 (bright) NK cells               1537   \n",
       "4694  CGGGCTTAGAGGAGTC-1       1     classical monocytes               1475   \n",
       "6716  GCATCCTTCAATCATG-1       1     CD56 (dim) NK cells               2520   \n",
       "484   AAGTGCAAGCCAAATC-1       1       naive CD4 T cells               1619   \n",
       "3087  CAATCCTGTCATAACG-1       1  CD56 (bright) NK cells               3081   \n",
       "5119  CTAGGACGTGTTCCCA-1       1         plasmacytoid DC               3108   \n",
       "1100  ACGCCACAGTAGCTTA-1       1          memory B cells               1907   \n",
       "7808  GGGTGTTGTTCGCTCA-1       1     classical monocytes               1874   \n",
       "\n",
       "      total_counts  total_counts_mt  pct_counts_mt  \n",
       "1506        2765.0              0.0            0.0  \n",
       "4670        4381.0              0.0            0.0  \n",
       "6010        2934.0              0.0            0.0  \n",
       "4694        2783.0              0.0            0.0  \n",
       "6716        6102.0              0.0            0.0  \n",
       "484         3398.0              0.0            0.0  \n",
       "3087        7406.0              0.0            0.0  \n",
       "5119        7025.0              0.0            0.0  \n",
       "1100        3958.0              0.0            0.0  \n",
       "7808        3686.0              0.0            0.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.obs.iloc[:10]            # metadata of 10 first cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4153b760",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>False</td>\n",
       "      <td>61</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>99.360788</td>\n",
       "      <td>66.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.2</th>\n",
       "      <td>AL627309.2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.5</th>\n",
       "      <td>AL627309.5</td>\n",
       "      <td>False</td>\n",
       "      <td>408</td>\n",
       "      <td>0.046002</td>\n",
       "      <td>95.724615</td>\n",
       "      <td>439.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.4</th>\n",
       "      <td>AL627309.4</td>\n",
       "      <td>False</td>\n",
       "      <td>41</td>\n",
       "      <td>0.004401</td>\n",
       "      <td>99.570366</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP006222.2</th>\n",
       "      <td>AP006222.2</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>99.989521</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL732372.1</th>\n",
       "      <td>AL732372.1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ids     mt  n_cells_by_counts  mean_counts  \\\n",
       "MIR1302-2HG  MIR1302-2HG  False                  0     0.000000   \n",
       "FAM138A          FAM138A  False                  0     0.000000   \n",
       "OR4F5              OR4F5  False                  0     0.000000   \n",
       "AL627309.1    AL627309.1  False                 61     0.006916   \n",
       "AL627309.3    AL627309.3  False                  0     0.000000   \n",
       "AL627309.2    AL627309.2  False                  0     0.000000   \n",
       "AL627309.5    AL627309.5  False                408     0.046002   \n",
       "AL627309.4    AL627309.4  False                 41     0.004401   \n",
       "AP006222.2    AP006222.2  False                  1     0.000105   \n",
       "AL732372.1    AL732372.1  False                  0     0.000000   \n",
       "\n",
       "             pct_dropout_by_counts  total_counts  \n",
       "MIR1302-2HG             100.000000           0.0  \n",
       "FAM138A                 100.000000           0.0  \n",
       "OR4F5                   100.000000           0.0  \n",
       "AL627309.1               99.360788          66.0  \n",
       "AL627309.3              100.000000           0.0  \n",
       "AL627309.2              100.000000           0.0  \n",
       "AL627309.5               95.724615         439.0  \n",
       "AL627309.4               99.570366          42.0  \n",
       "AP006222.2               99.989521           1.0  \n",
       "AL732372.1              100.000000           0.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.var.iloc[:10]            # metadata of 10 first genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d854fe0-e687-4516-add5-87d6f3ca20ba",
   "metadata": {},
   "source": [
    "## 5. Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dbf3f0d9-8d0a-4dba-ae8f-eb8433cf2daf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:139: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['n_genes'] = number\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:139: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['n_genes'] = number\n"
     ]
    }
   ],
   "source": [
    "# Filter low-count cells and genes\n",
    "\n",
    "# Keep only cells that have ≥ 200 detected genes\n",
    "sc.pp.filter_cells(adata_RNA, min_genes=200)\n",
    "\n",
    "# Keep only genes expressed in ≥ 3 cells\n",
    "sc.pp.filter_genes(adata_RNA, min_cells=3)\n",
    "\n",
    "sc.pp.filter_cells(adata_ATAC, min_genes=200)\n",
    "sc.pp.filter_genes(adata_ATAC, min_cells=3)\n",
    "\n",
    "# Keep only cells present in both RNA and ATAC\n",
    "selected_barcode = list(set(adata_RNA.obs['barcode'].values) & set(adata_ATAC.obs['barcode'].values))\n",
    "\n",
    "barcode_idx = pd.DataFrame(range(adata_RNA.shape[0]), index=adata_RNA.obs['barcode'].values)\n",
    "adata_RNA = adata_RNA[barcode_idx.loc[selected_barcode][0]]\n",
    "\n",
    "barcode_idx = pd.DataFrame(range(adata_ATAC.shape[0]), index=adata_ATAC.obs['barcode'].values)\n",
    "adata_ATAC = adata_ATAC[barcode_idx.loc[selected_barcode][0]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5da699d",
   "metadata": {},
   "source": [
    "### 5.1 Effect of preprocess\n",
    "We had 9543 cells and 144 978 features :\n",
    "- 36 601 genes\n",
    "- 108 377 peaks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "57fbc3f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adata_RNA.shape : (9543, 25485)\n",
      "adata_ATAC.shape : (9543, 107208)\n"
     ]
    }
   ],
   "source": [
    "print(f\"adata_RNA.shape : {adata_RNA.shape}\")\n",
    "print(f\"adata_ATAC.shape : {adata_ATAC.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7a99c35",
   "metadata": {},
   "source": [
    "### 5.2 Comparison with uncompressed data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e97b3f8",
   "metadata": {},
   "source": [
    "- Uncompressed : \n",
    "    - adata_RNA.shape : (9543, 36601)\n",
    "    - adata_ATAC.shape : (9543, 143887)\n",
    "- Compressed (.h5 file only)\n",
    "    - adata_RNA.shape : (9543, 36601)\n",
    "    - adata_ATAC.shape : (9543, 108377)\n",
    "\n",
    "**After preprocessing**\n",
    "\n",
    "- Uncompressed : \n",
    "    - adata_RNA.shape : (9543, 25485)\n",
    "    - adata_ATAC.shape : (9543, 143885)\n",
    "\n",
    "\n",
    "- Compressed (.h5 file only)\n",
    "    - adata_RNA.shape : (9543, 25485)\n",
    "    - adata_ATAC.shape : (9543, 107208)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a777f00-fdb5-4b9b-869d-57b5f98b3dae",
   "metadata": {},
   "source": [
    "### 5.3 About pseudo-bulking\n",
    "\n",
    "Pseudo-bulk means \"Combine many single cells into a “fake bulk sample” by summing or averaging their counts\"\n",
    "Since single cell is noisy and extremely sparse, it is better to work with aggregated signals across groups of cells (=metacells).  \n",
    "- ``singlepseudobulk = true`` : Collapse all cells in this sample into ONE pseudobulk profile. This gives following dimensions : \n",
    "    - TG_pseudobulk_temp : (n_genes × 1)\n",
    "    - RE_pseudobulk_temp : (n_peaks × 1) <br><br>\n",
    "      \n",
    "- ``singlepseudobulk = false`` : First cluster cells → then make multiple pseudobulks (metacells), used when we don't have many samples. This will create *K* clusters of cells, or *K* metacells\n",
    "    - TG_pseudobulk_temp : (n_genes × k_metacells)\n",
    "    - RE_pseudobulk_temp : (n_peaks × k_metacells) <br><br>\n",
    "      \n",
    "- Why is this needed ? GRN inference needs many samples (columns).\n",
    "    - If you already have many samples → 1 bulk per sample is enough\n",
    "    - If you have few samples → create metacells to increase sample count\n",
    "      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e7376eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_RNA[adata_RNA.obs['sample' ] == tempsample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb513ef2-844e-4187-a65d-045b4b79398b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/scanpy/preprocessing/_normalization.py:169: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:842: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:372: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:842: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "# Generate pseudo-bulk/metacell\n",
    "import os\n",
    "from LingerGRN.pseudo_bulk import *\n",
    "\n",
    "samplelist=list(set(adata_ATAC.obs['sample'].values)) # sample is generated from cell barcode \n",
    "tempsample=samplelist[0]\n",
    "\n",
    "TG_pseudobulk=pd.DataFrame([])\n",
    "RE_pseudobulk=pd.DataFrame([])\n",
    "\n",
    "n_samples = adata_RNA.obs['sample'].nunique()\n",
    "singlepseudobulk = (n_samples > 10)\n",
    "#singlepseudobulk = (adata_RNA.obs['sample'].unique().shape[0]*adata_RNA.obs['sample'].unique().shape[0]>100)\n",
    "\n",
    "# here samplelist = [1], singlepseudobulk = False (there is only one sample)\n",
    "for tempsample in samplelist:\n",
    "\n",
    "    # get cells from only tempsample\n",
    "    adata_RNAtemp = adata_RNA[adata_RNA.obs['sample' ] == tempsample]\n",
    "    adata_ATACtemp = adata_ATAC[adata_ATAC.obs['sample'] == tempsample]\n",
    "\n",
    "    TG_pseudobulk_temp, RE_pseudobulk_temp = pseudo_bulk(adata_RNAtemp, adata_ATACtemp, singlepseudobulk)  \n",
    "    \n",
    "    TG_pseudobulk = pd.concat([TG_pseudobulk, TG_pseudobulk_temp], axis=1)\n",
    "    RE_pseudobulk = pd.concat([RE_pseudobulk, RE_pseudobulk_temp], axis=1)\n",
    "    \n",
    "    RE_pseudobulk[RE_pseudobulk > 100] = 100\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d74f8606",
   "metadata": {},
   "source": [
    "### 5.4 About `pseudo_bulk` function\n",
    "\n",
    "**@inputs :**\n",
    "- adata_RNAtemp: single-cell RNA expression for one sample\n",
    "- adata_ATACtemp: single-cell chromatin accessibility for the same cells\n",
    "- singlepseudobulk (bool): whether to make one pseudo-bulk or multiple metacells\n",
    "\n",
    "**@outputs :**\n",
    "- TG_pseudobulk\n",
    "- RE_pseudobulk\n",
    "\n",
    "From one sample, it will either create a single metacell (single pseudobulk) or many pseudobulks. <br>\n",
    "Here, we will cluster 9,543 cells into 343 metacells (9543/343 ≃ 28 cells per metacell)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4f0cfe1e-8f21-4eab-9aa2-34077daf6562",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/anndata/_core/anndata.py:1230: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/ucl/inma/vangysel/miniconda3/envs/LINGER/lib/python3.10/site-packages/anndata/_core/anndata.py:1230: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists('data/'):\n",
    "    os.mkdir('data/')\n",
    "    \n",
    "adata_ATAC.write('data/adata_ATAC.h5ad')\n",
    "adata_RNA.write('data/adata_RNA.h5ad')\n",
    "\n",
    "TG_pseudobulk=TG_pseudobulk.fillna(0)\n",
    "RE_pseudobulk=RE_pseudobulk.fillna(0)\n",
    "\n",
    "pd.DataFrame(adata_ATAC.var['gene_ids']).to_csv('data/Peaks.txt',header=None,index=None)\n",
    "\n",
    "TG_pseudobulk.to_csv('data/TG_pseudobulk.tsv')\n",
    "RE_pseudobulk.to_csv('data/RE_pseudobulk.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d7a363ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCACCTAAGGAGCAAC-1</th>\n",
       "      <th>GTTACGTAGCTTTGGG-1</th>\n",
       "      <th>GGACTAAAGTCAATCA-1</th>\n",
       "      <th>AAACGGATCATGGCTG-1</th>\n",
       "      <th>CGCTAACCATTAGCCA-1</th>\n",
       "      <th>CGCATGATCGCTATGG-1</th>\n",
       "      <th>GGTAATTGTTTGTCTA-1</th>\n",
       "      <th>GTATGTTCAGAGGGAG-1</th>\n",
       "      <th>ACGTTACAGCCATCAG-1</th>\n",
       "      <th>ATGACGAAGTCACCAG-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TGCTTCCAGTAACCCG-1</th>\n",
       "      <th>CGGACCTAGCTAAAGG-1</th>\n",
       "      <th>CCTTAGTGTAGACAAA-1</th>\n",
       "      <th>CGCACAATCACCGGTA-1</th>\n",
       "      <th>CTAATCTTCTCGCCCA-1</th>\n",
       "      <th>CCCAACCGTAGCTGCG-1</th>\n",
       "      <th>AAGCGGGTCTTCAATC-1</th>\n",
       "      <th>CCTACTGGTTCGCGCT-1</th>\n",
       "      <th>GCTAAGAAGCTAAGTC-1</th>\n",
       "      <th>AAATGCCTCGATTTAG-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061358</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.112764</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.060099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282538</td>\n",
       "      <td>0.136926</td>\n",
       "      <td>0.185265</td>\n",
       "      <td>0.184987</td>\n",
       "      <td>0.119496</td>\n",
       "      <td>0.052138</td>\n",
       "      <td>0.075814</td>\n",
       "      <td>0.106592</td>\n",
       "      <td>0.146744</td>\n",
       "      <td>0.183203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065257</td>\n",
       "      <td>0.062803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.069169</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL669831.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01409</th>\n",
       "      <td>0.137802</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.073181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103896</td>\n",
       "      <td>0.144276</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.067568</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.164827</td>\n",
       "      <td>0.109030</td>\n",
       "      <td>0.060882</td>\n",
       "      <td>0.058678</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.073396</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL592183.1</th>\n",
       "      <td>0.146648</td>\n",
       "      <td>0.325133</td>\n",
       "      <td>0.072462</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.247954</td>\n",
       "      <td>0.205121</td>\n",
       "      <td>0.147658</td>\n",
       "      <td>0.138384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133816</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.143415</td>\n",
       "      <td>0.062079</td>\n",
       "      <td>0.057773</td>\n",
       "      <td>0.151881</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.073540</td>\n",
       "      <td>0.056966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.125495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>0.072462</td>\n",
       "      <td>0.171601</td>\n",
       "      <td>0.072462</td>\n",
       "      <td>0.067232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.067232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.067232</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.063373</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072813</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.067697</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004556.3</th>\n",
       "      <td>0.076179</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076179</td>\n",
       "      <td>0.063236</td>\n",
       "      <td>0.072878</td>\n",
       "      <td>0.105052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.131637</td>\n",
       "      <td>0.082473</td>\n",
       "      <td>0.209778</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.053963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103969</td>\n",
       "      <td>0.073348</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.064783</td>\n",
       "      <td>0.064783</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.052947</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25485 rows × 343 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GCACCTAAGGAGCAAC-1  GTTACGTAGCTTTGGG-1  GGACTAAAGTCAATCA-1  \\\n",
       "AL627309.1            0.000000            0.000000            0.000000   \n",
       "AL627309.5            0.000000            0.000000            0.000000   \n",
       "AL627309.4            0.000000            0.000000            0.000000   \n",
       "AL669831.2            0.000000            0.000000            0.000000   \n",
       "LINC01409             0.137802            0.000000            0.073181   \n",
       "...                        ...                 ...                 ...   \n",
       "AL592183.1            0.146648            0.325133            0.072462   \n",
       "AC240274.1            0.072462            0.171601            0.072462   \n",
       "AC004556.3            0.076179            0.000000            0.076179   \n",
       "AC007325.1            0.000000            0.000000            0.000000   \n",
       "AC007325.4            0.000000            0.000000            0.000000   \n",
       "\n",
       "            AAACGGATCATGGCTG-1  CGCTAACCATTAGCCA-1  CGCATGATCGCTATGG-1  \\\n",
       "AL627309.1            0.000000            0.000000            0.000000   \n",
       "AL627309.5            0.000000            0.060099            0.000000   \n",
       "AL627309.4            0.000000            0.000000            0.000000   \n",
       "AL669831.2            0.000000            0.000000            0.000000   \n",
       "LINC01409             0.000000            0.103896            0.144276   \n",
       "...                        ...                 ...                 ...   \n",
       "AL592183.1            0.000000            0.247954            0.205121   \n",
       "AC240274.1            0.067232            0.000000            0.000000   \n",
       "AC004556.3            0.063236            0.072878            0.105052   \n",
       "AC007325.1            0.000000            0.000000            0.000000   \n",
       "AC007325.4            0.000000            0.000000            0.000000   \n",
       "\n",
       "            GGTAATTGTTTGTCTA-1  GTATGTTCAGAGGGAG-1  ACGTTACAGCCATCAG-1  \\\n",
       "AL627309.1            0.000000            0.000000            0.000000   \n",
       "AL627309.5            0.000000            0.000000            0.000000   \n",
       "AL627309.4            0.000000            0.000000            0.000000   \n",
       "AL669831.2            0.000000            0.000000            0.000000   \n",
       "LINC01409             0.000000            0.067568            0.000000   \n",
       "...                        ...                 ...                 ...   \n",
       "AL592183.1            0.147658            0.138384            0.000000   \n",
       "AC240274.1            0.000000            0.067232            0.000000   \n",
       "AC004556.3            0.000000            0.131637            0.082473   \n",
       "AC007325.1            0.000000            0.000000            0.000000   \n",
       "AC007325.4            0.000000            0.000000            0.064783   \n",
       "\n",
       "            ATGACGAAGTCACCAG-1  ...  TGCTTCCAGTAACCCG-1  CGGACCTAGCTAAAGG-1  \\\n",
       "AL627309.1            0.000000  ...            0.000000            0.000000   \n",
       "AL627309.5            0.000000  ...            0.282538            0.136926   \n",
       "AL627309.4            0.000000  ...            0.000000            0.065257   \n",
       "AL669831.2            0.000000  ...            0.000000            0.000000   \n",
       "LINC01409             0.000000  ...            0.000000            0.164827   \n",
       "...                        ...  ...                 ...                 ...   \n",
       "AL592183.1            0.133816  ...            0.000000            0.143415   \n",
       "AC240274.1            0.067232  ...            0.000000            0.063373   \n",
       "AC004556.3            0.209778  ...            0.000000            0.053963   \n",
       "AC007325.1            0.000000  ...            0.000000            0.000000   \n",
       "AC007325.4            0.064783  ...            0.000000            0.000000   \n",
       "\n",
       "            CCTTAGTGTAGACAAA-1  CGCACAATCACCGGTA-1  CTAATCTTCTCGCCCA-1  \\\n",
       "AL627309.1            0.000000            0.061358            0.000000   \n",
       "AL627309.5            0.185265            0.184987            0.119496   \n",
       "AL627309.4            0.062803            0.000000            0.000000   \n",
       "AL669831.2            0.000000            0.000000            0.000000   \n",
       "LINC01409             0.109030            0.060882            0.058678   \n",
       "...                        ...                 ...                 ...   \n",
       "AL592183.1            0.062079            0.057773            0.151881   \n",
       "AC240274.1            0.000000            0.072813            0.000000   \n",
       "AC004556.3            0.000000            0.000000            0.000000   \n",
       "AC007325.1            0.000000            0.000000            0.000000   \n",
       "AC007325.4            0.000000            0.000000            0.000000   \n",
       "\n",
       "            CCCAACCGTAGCTGCG-1  AAGCGGGTCTTCAATC-1  CCTACTGGTTCGCGCT-1  \\\n",
       "AL627309.1            0.000000            0.112764            0.000000   \n",
       "AL627309.5            0.052138            0.075814            0.106592   \n",
       "AL627309.4            0.000000            0.069169            0.000000   \n",
       "AL669831.2            0.000000            0.000000            0.000000   \n",
       "LINC01409             0.000000            0.000000            0.000000   \n",
       "...                        ...                 ...                 ...   \n",
       "AL592183.1            0.000000            0.073540            0.056966   \n",
       "AC240274.1            0.000000            0.000000            0.000000   \n",
       "AC004556.3            0.000000            0.000000            0.103969   \n",
       "AC007325.1            0.000000            0.000000            0.000000   \n",
       "AC007325.4            0.000000            0.000000            0.052947   \n",
       "\n",
       "            GCTAAGAAGCTAAGTC-1  AAATGCCTCGATTTAG-1  \n",
       "AL627309.1            0.000000            0.000000  \n",
       "AL627309.5            0.146744            0.183203  \n",
       "AL627309.4            0.000000            0.000000  \n",
       "AL669831.2            0.000000            0.000000  \n",
       "LINC01409             0.073396            0.000000  \n",
       "...                        ...                 ...  \n",
       "AL592183.1            0.000000            0.125495  \n",
       "AC240274.1            0.067697            0.000000  \n",
       "AC004556.3            0.073348            0.000000  \n",
       "AC007325.1            0.000000            0.000000  \n",
       "AC007325.4            0.000000            0.000000  \n",
       "\n",
       "[25485 rows x 343 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TG_pseudobulk        # 25 485 genes x 343 bulks (meta cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a90a5886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCACCTAAGGAGCAAC-1</th>\n",
       "      <th>GTTACGTAGCTTTGGG-1</th>\n",
       "      <th>GGACTAAAGTCAATCA-1</th>\n",
       "      <th>AAACGGATCATGGCTG-1</th>\n",
       "      <th>CGCTAACCATTAGCCA-1</th>\n",
       "      <th>CGCATGATCGCTATGG-1</th>\n",
       "      <th>GGTAATTGTTTGTCTA-1</th>\n",
       "      <th>GTATGTTCAGAGGGAG-1</th>\n",
       "      <th>ACGTTACAGCCATCAG-1</th>\n",
       "      <th>ATGACGAAGTCACCAG-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TGCTTCCAGTAACCCG-1</th>\n",
       "      <th>CGGACCTAGCTAAAGG-1</th>\n",
       "      <th>CCTTAGTGTAGACAAA-1</th>\n",
       "      <th>CGCACAATCACCGGTA-1</th>\n",
       "      <th>CTAATCTTCTCGCCCA-1</th>\n",
       "      <th>CCCAACCGTAGCTGCG-1</th>\n",
       "      <th>AAGCGGGTCTTCAATC-1</th>\n",
       "      <th>CCTACTGGTTCGCGCT-1</th>\n",
       "      <th>GCTAAGAAGCTAAGTC-1</th>\n",
       "      <th>AAATGCCTCGATTTAG-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:10109-10357</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:180730-181630</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:191491-191736</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:267816-268196</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.084707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:586028-586373</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:20444-22615</th>\n",
       "      <td>0.445296</td>\n",
       "      <td>0.431637</td>\n",
       "      <td>0.815061</td>\n",
       "      <td>0.458523</td>\n",
       "      <td>0.404752</td>\n",
       "      <td>0.427587</td>\n",
       "      <td>0.209947</td>\n",
       "      <td>0.574166</td>\n",
       "      <td>0.485409</td>\n",
       "      <td>0.373816</td>\n",
       "      <td>...</td>\n",
       "      <td>0.495004</td>\n",
       "      <td>0.707519</td>\n",
       "      <td>0.823594</td>\n",
       "      <td>0.832770</td>\n",
       "      <td>0.246428</td>\n",
       "      <td>0.587825</td>\n",
       "      <td>0.530003</td>\n",
       "      <td>0.618761</td>\n",
       "      <td>0.837884</td>\n",
       "      <td>0.88611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:27118-28927</th>\n",
       "      <td>0.130785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.130785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.084707</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>...</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.115643</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:29485-30706</th>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072963</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:31511-32072</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:37129-37638</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107208 rows × 343 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GCACCTAAGGAGCAAC-1  GTTACGTAGCTTTGGG-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.000000            0.000000   \n",
       "chr1:191491-191736                0.000000            0.000000   \n",
       "chr1:267816-268196                0.000000            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.445296            0.431637   \n",
       "KI270713.1:27118-28927            0.130785            0.000000   \n",
       "KI270713.1:29485-30706            0.057822            0.057822   \n",
       "KI270713.1:31511-32072            0.000000            0.000000   \n",
       "KI270713.1:37129-37638            0.000000            0.000000   \n",
       "\n",
       "                        GGACTAAAGTCAATCA-1  AAACGGATCATGGCTG-1  \\\n",
       "chr1:10109-10357                  0.000000            0.057822   \n",
       "chr1:180730-181630                0.000000            0.000000   \n",
       "chr1:191491-191736                0.000000            0.000000   \n",
       "chr1:267816-268196                0.000000            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.815061            0.458523   \n",
       "KI270713.1:27118-28927            0.130785            0.000000   \n",
       "KI270713.1:29485-30706            0.057822            0.000000   \n",
       "KI270713.1:31511-32072            0.000000            0.000000   \n",
       "KI270713.1:37129-37638            0.000000            0.000000   \n",
       "\n",
       "                        CGCTAACCATTAGCCA-1  CGCATGATCGCTATGG-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.057822            0.036481   \n",
       "chr1:191491-191736                0.057822            0.000000   \n",
       "chr1:267816-268196                0.057822            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.404752            0.427587   \n",
       "KI270713.1:27118-28927            0.084707            0.036481   \n",
       "KI270713.1:29485-30706            0.057822            0.057822   \n",
       "KI270713.1:31511-32072            0.000000            0.000000   \n",
       "KI270713.1:37129-37638            0.000000            0.000000   \n",
       "\n",
       "                        GGTAATTGTTTGTCTA-1  GTATGTTCAGAGGGAG-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.000000            0.057822   \n",
       "chr1:191491-191736                0.000000            0.000000   \n",
       "chr1:267816-268196                0.000000            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.209947            0.574166   \n",
       "KI270713.1:27118-28927            0.000000            0.057822   \n",
       "KI270713.1:29485-30706            0.000000            0.000000   \n",
       "KI270713.1:31511-32072            0.000000            0.057822   \n",
       "KI270713.1:37129-37638            0.000000            0.000000   \n",
       "\n",
       "                        ACGTTACAGCCATCAG-1  ATGACGAAGTCACCAG-1  ...  \\\n",
       "chr1:10109-10357                  0.000000            0.000000  ...   \n",
       "chr1:180730-181630                0.000000            0.000000  ...   \n",
       "chr1:191491-191736                0.000000            0.000000  ...   \n",
       "chr1:267816-268196                0.036481            0.000000  ...   \n",
       "chr1:586028-586373                0.000000            0.000000  ...   \n",
       "...                                    ...                 ...  ...   \n",
       "KI270713.1:20444-22615            0.485409            0.373816  ...   \n",
       "KI270713.1:27118-28927            0.057822            0.057822  ...   \n",
       "KI270713.1:29485-30706            0.000000            0.000000  ...   \n",
       "KI270713.1:31511-32072            0.000000            0.000000  ...   \n",
       "KI270713.1:37129-37638            0.000000            0.000000  ...   \n",
       "\n",
       "                        TGCTTCCAGTAACCCG-1  CGGACCTAGCTAAAGG-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.000000            0.000000   \n",
       "chr1:191491-191736                0.000000            0.000000   \n",
       "chr1:267816-268196                0.057822            0.084707   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.495004            0.707519   \n",
       "KI270713.1:27118-28927            0.036481            0.000000   \n",
       "KI270713.1:29485-30706            0.000000            0.000000   \n",
       "KI270713.1:31511-32072            0.000000            0.000000   \n",
       "KI270713.1:37129-37638            0.000000            0.000000   \n",
       "\n",
       "                        CCTTAGTGTAGACAAA-1  CGCACAATCACCGGTA-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.057822            0.036481   \n",
       "chr1:191491-191736                0.000000            0.000000   \n",
       "chr1:267816-268196                0.000000            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.823594            0.832770   \n",
       "KI270713.1:27118-28927            0.000000            0.000000   \n",
       "KI270713.1:29485-30706            0.057822            0.000000   \n",
       "KI270713.1:31511-32072            0.000000            0.000000   \n",
       "KI270713.1:37129-37638            0.057822            0.000000   \n",
       "\n",
       "                        CTAATCTTCTCGCCCA-1  CCCAACCGTAGCTGCG-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.000000            0.000000   \n",
       "chr1:191491-191736                0.000000            0.036481   \n",
       "chr1:267816-268196                0.000000            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.246428            0.587825   \n",
       "KI270713.1:27118-28927            0.000000            0.057822   \n",
       "KI270713.1:29485-30706            0.000000            0.057822   \n",
       "KI270713.1:31511-32072            0.000000            0.000000   \n",
       "KI270713.1:37129-37638            0.000000            0.036481   \n",
       "\n",
       "                        AAGCGGGTCTTCAATC-1  CCTACTGGTTCGCGCT-1  \\\n",
       "chr1:10109-10357                  0.000000            0.000000   \n",
       "chr1:180730-181630                0.000000            0.000000   \n",
       "chr1:191491-191736                0.000000            0.000000   \n",
       "chr1:267816-268196                0.000000            0.000000   \n",
       "chr1:586028-586373                0.000000            0.000000   \n",
       "...                                    ...                 ...   \n",
       "KI270713.1:20444-22615            0.530003            0.618761   \n",
       "KI270713.1:27118-28927            0.057822            0.057822   \n",
       "KI270713.1:29485-30706            0.057822            0.000000   \n",
       "KI270713.1:31511-32072            0.057822            0.000000   \n",
       "KI270713.1:37129-37638            0.000000            0.057822   \n",
       "\n",
       "                        GCTAAGAAGCTAAGTC-1  AAATGCCTCGATTTAG-1  \n",
       "chr1:10109-10357                  0.000000             0.00000  \n",
       "chr1:180730-181630                0.000000             0.00000  \n",
       "chr1:191491-191736                0.036481             0.00000  \n",
       "chr1:267816-268196                0.000000             0.00000  \n",
       "chr1:586028-586373                0.000000             0.00000  \n",
       "...                                    ...                 ...  \n",
       "KI270713.1:20444-22615            0.837884             0.88611  \n",
       "KI270713.1:27118-28927            0.115643             0.00000  \n",
       "KI270713.1:29485-30706            0.072963             0.00000  \n",
       "KI270713.1:31511-32072            0.000000             0.00000  \n",
       "KI270713.1:37129-37638            0.057822             0.00000  \n",
       "\n",
       "[107208 rows x 343 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RE_pseudobulk         # 107 208 peaks x 343 bulks "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "38c5fc47-277e-4bc1-ba64-f1c8d63a60cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/users/v/a/vangysel/linger\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e1ecb67-a0f9-4d67-85aa-1137647edb91",
   "metadata": {},
   "source": [
    "## 6. Training the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "da174406-b2c7-4838-968c-6b00263e23f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from LingerGRN.preprocess import *\n",
    "\n",
    "#Datadir = os.path.join(os.getcwd(), 'LINGER_data/')\n",
    "Datadir = \"/globalscratch/ucl/inma/vangysel/\" + \"LINGER_data/\"\n",
    "GRNdir = Datadir + 'data_bulk/'\n",
    "genome = 'hg38'\n",
    "#outdir = '/LINGER_output/'  # output directory\n",
    "outdir = \"/globalscratch/ucl/inma/vangysel/\" + \"LINGER_output/\"\n",
    "method = 'baseline'         # or 'LINGER'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a88b23a0-ec2a-4a23-ac72-c4a7273f5f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlap the regions with bulk data ...\n"
     ]
    }
   ],
   "source": [
    "preprocess(TG_pseudobulk, RE_pseudobulk, GRNdir, genome, method, outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "49aa6e06-5199-4db0-8ebb-a42143e31041",
   "metadata": {},
   "outputs": [],
   "source": [
    "import LingerGRN.LINGER_tr as LINGER_tr\n",
    "\n",
    "activef='ReLU' # active function chose from 'ReLU','sigmoid','tanh'\n",
    "LINGER_tr.training(GRNdir,method,outdir,activef,'Human')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7928d8b-8937-4b55-ae40-9d1570089d17",
   "metadata": {},
   "source": [
    "## 7. Cell population gene regulatory network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5df7c578-3725-45e8-8109-5be5c652a518",
   "metadata": {},
   "source": [
    "### 7.1 TF binding potential (TF-RE)\n",
    "The output is 'cell_population_TF_RE_binding.txt', a matrix of the TF-RE binding score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d39a68e4-b60a-4720-a73b-062d014c7cac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating cellular population TF binding strength ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████| 23/23 [00:57<00:00,  2.49s/it]\n"
     ]
    }
   ],
   "source": [
    "import LingerGRN.LL_net as LL_net\n",
    "LL_net.TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "711f0ac7-c653-4951-be17-7ba222f13d66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZBTB6</th>\n",
       "      <th>ARNTL2</th>\n",
       "      <th>ELK3</th>\n",
       "      <th>E2F1</th>\n",
       "      <th>TCF7</th>\n",
       "      <th>TCF4</th>\n",
       "      <th>JUND</th>\n",
       "      <th>ZSCAN22</th>\n",
       "      <th>MEF2D</th>\n",
       "      <th>HLX</th>\n",
       "      <th>...</th>\n",
       "      <th>TBX21</th>\n",
       "      <th>MZF1</th>\n",
       "      <th>LHX4</th>\n",
       "      <th>KLF16</th>\n",
       "      <th>NFIC</th>\n",
       "      <th>FOXJ3</th>\n",
       "      <th>POU6F2</th>\n",
       "      <th>SIX3</th>\n",
       "      <th>ZIC1</th>\n",
       "      <th>CENPB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.274914</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.471285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.577039</td>\n",
       "      <td>0.558025</td>\n",
       "      <td>0.071134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.855215</td>\n",
       "      <td>0.406899</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.400902</td>\n",
       "      <td>0.062382</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.541846</td>\n",
       "      <td>0.341378</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.269262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100034436-100035279</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100035922-100040109</th>\n",
       "      <td>0.174891</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.504716</td>\n",
       "      <td>0.517765</td>\n",
       "      <td>0.117817</td>\n",
       "      <td>0.194641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.196663</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.724422</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.077789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100041493-100041927</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100046068-100047735</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.548376</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.623037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.375218</td>\n",
       "      <td>0.350451</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.570573</td>\n",
       "      <td>0.493897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.278602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.382428</td>\n",
       "      <td>0.517982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.073065</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 451 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             ZBTB6    ARNTL2      ELK3      E2F1      TCF7  \\\n",
       "chr1:100028489-100029404  0.000000  0.274914  0.000000  0.000000  0.471285   \n",
       "chr1:100034436-100035279  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "chr1:100035922-100040109  0.174891  0.000000  0.504716  0.517765  0.117817   \n",
       "chr1:100041493-100041927  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "chr1:100046068-100047735  0.000000  0.548376  0.000000  0.000000  0.623037   \n",
       "\n",
       "                              TCF4      JUND   ZSCAN22     MEF2D       HLX  \\\n",
       "chr1:100028489-100029404  0.000000  0.577039  0.558025  0.071134  0.000000   \n",
       "chr1:100034436-100035279  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "chr1:100035922-100040109  0.194641  0.000000  0.000000  0.000000  0.196663   \n",
       "chr1:100041493-100041927  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "chr1:100046068-100047735  0.000000  0.375218  0.350451  0.000000  0.000000   \n",
       "\n",
       "                          ...     TBX21      MZF1      LHX4     KLF16  \\\n",
       "chr1:100028489-100029404  ...  0.855215  0.406899  0.000000  0.400902   \n",
       "chr1:100034436-100035279  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "chr1:100035922-100040109  ...  0.000000  0.000000  0.724422  0.000000   \n",
       "chr1:100041493-100041927  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "chr1:100046068-100047735  ...  0.570573  0.493897  0.000000  0.278602   \n",
       "\n",
       "                              NFIC     FOXJ3    POU6F2      SIX3  ZIC1  \\\n",
       "chr1:100028489-100029404  0.062382  0.000000  0.541846  0.341378   0.0   \n",
       "chr1:100034436-100035279  0.000000  0.000000  0.000000  0.000000   0.0   \n",
       "chr1:100035922-100040109  0.000000  0.077789  0.000000  0.000000   0.0   \n",
       "chr1:100041493-100041927  0.000000  0.000000  0.000000  0.000000   0.0   \n",
       "chr1:100046068-100047735  0.000000  0.000000  0.382428  0.517982   0.0   \n",
       "\n",
       "                             CENPB  \n",
       "chr1:100028489-100029404  0.269262  \n",
       "chr1:100034436-100035279  0.000000  \n",
       "chr1:100035922-100040109  0.000000  \n",
       "chr1:100041493-100041927  0.000000  \n",
       "chr1:100046068-100047735  0.073065  \n",
       "\n",
       "[5 rows x 451 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_re = pd.read_csv(\"/globalscratch/ucl/inma/vangysel/LINGER_output/cell_population_TF_RE_binding.txt\", sep=\"\\t\", index_col=0)\n",
    "tf_re.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab01b819-658e-4833-be5d-4a08d921cc49",
   "metadata": {},
   "source": [
    "### 7.2 cis-regulatory network (RE-TG)\n",
    "The output is 'cell_population_cis_regulatory.txt' with 3 columns: region, target gene, cis-regulatory score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7a412ca2-5ba5-4130-80e2-bd14486f82ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████| 23/23 [02:09<00:00,  5.65s/it]\n"
     ]
    }
   ],
   "source": [
    "LL_net.cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c48e2019-ea2a-4dd1-a48a-77a04d786e01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SASS6</th>\n",
       "      <th>0.0009586007532124838</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <td>RTCA</td>\n",
       "      <td>4.597646e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <td>SLC30A7</td>\n",
       "      <td>5.400717e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <td>LRRC39</td>\n",
       "      <td>1.534570e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <td>CDC14A</td>\n",
       "      <td>2.005422e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:100028489-100029404</th>\n",
       "      <td>FRRS1</td>\n",
       "      <td>1.796895e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            SASS6  0.0009586007532124838\n",
       "chr1:100028489-100029404                                \n",
       "chr1:100028489-100029404     RTCA           4.597646e-06\n",
       "chr1:100028489-100029404  SLC30A7           5.400717e-17\n",
       "chr1:100028489-100029404   LRRC39           1.534570e-04\n",
       "chr1:100028489-100029404   CDC14A           2.005422e-07\n",
       "chr1:100028489-100029404    FRRS1           1.796895e-06"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re_tg = pd.read_csv(\"/globalscratch/ucl/inma/vangysel/LINGER_output/cell_population_cis_regulatory.txt\", sep=\"\\t\", index_col=0)\n",
    "re_tg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc94a2a-74a1-4062-bf10-26c6d14caae8",
   "metadata": {},
   "source": [
    "### 7.3 trans-regulatory network (TF-TG)\n",
    "The output is 'cell_population_trans_regulatory.txt', a matrix of the trans-regulatory score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "52206b73-71b7-44bb-9ee6-999e79528bef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generate trans-regulatory netowrk ...\n",
      "Save trans-regulatory netowrk ...\n"
     ]
    }
   ],
   "source": [
    "LL_net.trans_reg(GRNdir,method,outdir,genome)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b0323118-d6fb-47a9-8936-dbc88028fa99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZBTB6</th>\n",
       "      <th>ARNTL2</th>\n",
       "      <th>ELK3</th>\n",
       "      <th>E2F1</th>\n",
       "      <th>TCF7</th>\n",
       "      <th>TCF4</th>\n",
       "      <th>JUND</th>\n",
       "      <th>ZSCAN22</th>\n",
       "      <th>MEF2D</th>\n",
       "      <th>HLX</th>\n",
       "      <th>...</th>\n",
       "      <th>TBX21</th>\n",
       "      <th>MZF1</th>\n",
       "      <th>LHX4</th>\n",
       "      <th>KLF16</th>\n",
       "      <th>NFIC</th>\n",
       "      <th>FOXJ3</th>\n",
       "      <th>POU6F2</th>\n",
       "      <th>SIX3</th>\n",
       "      <th>ZIC1</th>\n",
       "      <th>CENPB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SASS6</th>\n",
       "      <td>1.594028e-06</td>\n",
       "      <td>6.235565e-05</td>\n",
       "      <td>6.701126e-05</td>\n",
       "      <td>1.361343e-04</td>\n",
       "      <td>9.386463e-05</td>\n",
       "      <td>7.419297e-07</td>\n",
       "      <td>1.186786e-05</td>\n",
       "      <td>6.355007e-06</td>\n",
       "      <td>4.090155e-07</td>\n",
       "      <td>1.937167e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>5.155972e-06</td>\n",
       "      <td>6.182847e-05</td>\n",
       "      <td>1.224008e-04</td>\n",
       "      <td>4.554059e-05</td>\n",
       "      <td>8.231278e-08</td>\n",
       "      <td>1.481097e-07</td>\n",
       "      <td>5.551017e-05</td>\n",
       "      <td>1.111126e-05</td>\n",
       "      <td>2.872886e-08</td>\n",
       "      <td>2.529586e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RTCA</th>\n",
       "      <td>7.793802e-06</td>\n",
       "      <td>7.384969e-05</td>\n",
       "      <td>7.840596e-05</td>\n",
       "      <td>8.340468e-05</td>\n",
       "      <td>2.640587e-04</td>\n",
       "      <td>4.488400e-06</td>\n",
       "      <td>7.010572e-05</td>\n",
       "      <td>2.946676e-05</td>\n",
       "      <td>1.266470e-05</td>\n",
       "      <td>1.264958e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>4.269316e-05</td>\n",
       "      <td>1.927490e-04</td>\n",
       "      <td>1.339696e-04</td>\n",
       "      <td>1.422249e-04</td>\n",
       "      <td>2.043960e-05</td>\n",
       "      <td>1.231631e-06</td>\n",
       "      <td>8.481464e-05</td>\n",
       "      <td>2.723950e-05</td>\n",
       "      <td>5.167986e-06</td>\n",
       "      <td>1.797284e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC30A7</th>\n",
       "      <td>1.521859e-08</td>\n",
       "      <td>9.781391e-09</td>\n",
       "      <td>1.062280e-08</td>\n",
       "      <td>3.638231e-09</td>\n",
       "      <td>2.140661e-08</td>\n",
       "      <td>1.626306e-08</td>\n",
       "      <td>1.520017e-07</td>\n",
       "      <td>1.719834e-07</td>\n",
       "      <td>7.796982e-08</td>\n",
       "      <td>1.081337e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>1.019648e-07</td>\n",
       "      <td>3.065578e-08</td>\n",
       "      <td>8.333568e-09</td>\n",
       "      <td>9.603236e-08</td>\n",
       "      <td>1.817609e-09</td>\n",
       "      <td>3.691107e-08</td>\n",
       "      <td>4.675161e-08</td>\n",
       "      <td>8.175803e-09</td>\n",
       "      <td>1.684074e-08</td>\n",
       "      <td>2.526880e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRRC39</th>\n",
       "      <td>5.558691e-07</td>\n",
       "      <td>2.173287e-05</td>\n",
       "      <td>7.911914e-06</td>\n",
       "      <td>1.112355e-05</td>\n",
       "      <td>2.148481e-05</td>\n",
       "      <td>3.513813e-07</td>\n",
       "      <td>3.777721e-06</td>\n",
       "      <td>1.121292e-06</td>\n",
       "      <td>3.102803e-07</td>\n",
       "      <td>1.812656e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>2.669588e-06</td>\n",
       "      <td>2.063336e-05</td>\n",
       "      <td>2.690603e-05</td>\n",
       "      <td>5.479154e-06</td>\n",
       "      <td>2.288930e-07</td>\n",
       "      <td>1.020716e-07</td>\n",
       "      <td>1.052551e-05</td>\n",
       "      <td>8.968327e-07</td>\n",
       "      <td>1.346092e-07</td>\n",
       "      <td>7.692850e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDC14A</th>\n",
       "      <td>3.402137e-05</td>\n",
       "      <td>2.014284e-04</td>\n",
       "      <td>1.970612e-04</td>\n",
       "      <td>2.280473e-04</td>\n",
       "      <td>9.994210e-05</td>\n",
       "      <td>4.465963e-05</td>\n",
       "      <td>2.470534e-04</td>\n",
       "      <td>1.122330e-04</td>\n",
       "      <td>1.421436e-04</td>\n",
       "      <td>1.335638e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>1.932471e-04</td>\n",
       "      <td>1.754173e-04</td>\n",
       "      <td>1.120573e-04</td>\n",
       "      <td>1.842412e-04</td>\n",
       "      <td>2.385806e-04</td>\n",
       "      <td>5.179631e-05</td>\n",
       "      <td>1.173447e-04</td>\n",
       "      <td>1.260149e-05</td>\n",
       "      <td>7.508897e-05</td>\n",
       "      <td>4.078610e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 451 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ZBTB6        ARNTL2          ELK3          E2F1          TCF7  \\\n",
       "SASS6    1.594028e-06  6.235565e-05  6.701126e-05  1.361343e-04  9.386463e-05   \n",
       "RTCA     7.793802e-06  7.384969e-05  7.840596e-05  8.340468e-05  2.640587e-04   \n",
       "SLC30A7  1.521859e-08  9.781391e-09  1.062280e-08  3.638231e-09  2.140661e-08   \n",
       "LRRC39   5.558691e-07  2.173287e-05  7.911914e-06  1.112355e-05  2.148481e-05   \n",
       "CDC14A   3.402137e-05  2.014284e-04  1.970612e-04  2.280473e-04  9.994210e-05   \n",
       "\n",
       "                 TCF4          JUND       ZSCAN22         MEF2D           HLX  \\\n",
       "SASS6    7.419297e-07  1.186786e-05  6.355007e-06  4.090155e-07  1.937167e-07   \n",
       "RTCA     4.488400e-06  7.010572e-05  2.946676e-05  1.266470e-05  1.264958e-06   \n",
       "SLC30A7  1.626306e-08  1.520017e-07  1.719834e-07  7.796982e-08  1.081337e-08   \n",
       "LRRC39   3.513813e-07  3.777721e-06  1.121292e-06  3.102803e-07  1.812656e-07   \n",
       "CDC14A   4.465963e-05  2.470534e-04  1.122330e-04  1.421436e-04  1.335638e-05   \n",
       "\n",
       "         ...         TBX21          MZF1          LHX4         KLF16  \\\n",
       "SASS6    ...  5.155972e-06  6.182847e-05  1.224008e-04  4.554059e-05   \n",
       "RTCA     ...  4.269316e-05  1.927490e-04  1.339696e-04  1.422249e-04   \n",
       "SLC30A7  ...  1.019648e-07  3.065578e-08  8.333568e-09  9.603236e-08   \n",
       "LRRC39   ...  2.669588e-06  2.063336e-05  2.690603e-05  5.479154e-06   \n",
       "CDC14A   ...  1.932471e-04  1.754173e-04  1.120573e-04  1.842412e-04   \n",
       "\n",
       "                 NFIC         FOXJ3        POU6F2          SIX3          ZIC1  \\\n",
       "SASS6    8.231278e-08  1.481097e-07  5.551017e-05  1.111126e-05  2.872886e-08   \n",
       "RTCA     2.043960e-05  1.231631e-06  8.481464e-05  2.723950e-05  5.167986e-06   \n",
       "SLC30A7  1.817609e-09  3.691107e-08  4.675161e-08  8.175803e-09  1.684074e-08   \n",
       "LRRC39   2.288930e-07  1.020716e-07  1.052551e-05  8.968327e-07  1.346092e-07   \n",
       "CDC14A   2.385806e-04  5.179631e-05  1.173447e-04  1.260149e-05  7.508897e-05   \n",
       "\n",
       "                CENPB  \n",
       "SASS6    2.529586e-06  \n",
       "RTCA     1.797284e-05  \n",
       "SLC30A7  2.526880e-08  \n",
       "LRRC39   7.692850e-07  \n",
       "CDC14A   4.078610e-04  \n",
       "\n",
       "[5 rows x 451 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_tg = pd.read_csv(\"/globalscratch/ucl/inma/vangysel/LINGER_output/cell_population_trans_regulatory.txt\", sep=\"\\t\", index_col=0)\n",
    "tf_tg.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (LINGER)",
   "language": "python",
   "name": "linger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
