{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcacd40f-2c6a-49de-b3d1-ed26c9103f24",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1ec52e13",
   "metadata": {},
   "source": [
    "# PBMCs Tutorial\n",
    "PBMC stands for Peripheral Blood Mononuclear Cells, a vital group of immune system cells found in the blood\n",
    "\n",
    "## Instructions\n",
    "This tutorial delineates a computational framework for constructing gene regulatory networks (GRNs) from single-cell multiome data. We provide 2 options to do this: 'baseline' and 'LINGER'. \n",
    "The first is a naive method combining the prior GRNs and features from the single-cell data, offering a rapid approach. \n",
    "LINGER integrates the comprehensive gene regulatory profile from external bulk data. As the following figure, LINGER uses lifelong machine learning (continuous learning) based on neural network (NN) models, which has been proven to leverage the knowledge learned in previous tasks to help learn the new task better. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1488580d-c7eb-4f5e-b2ea-28de4670d29a",
   "metadata": {},
   "source": [
    "## 1. Download the general gene regulatory network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ea0e4a80-49c7-4c14-970b-e259179737f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/emile/Documents/UCL/Memoire/Lingier\n"
     ]
    }
   ],
   "source": [
    "!echo $(pwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab75d5f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/emile/Documents/UCL/Memoire/Lingier/LINGER_data\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Set directories and download general GRN\n",
    "Datadir=$PWD/LINGER_data\n",
    "mkdir -p $Datadir\n",
    "cd $Datadir\n",
    "echo $(pwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfa9da48-7a14-48f0-9d42-ff1c06d5bd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Download general GRN from Google Drive\n",
    "wget --load-cookies /tmp/cookies.txt \"https://drive.usercontent.google.com/download?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://drive.usercontent.google.com/download?id=1jwRgRHPJrKABOk7wImKONTtUupV7yJ9b'  -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\\1\\n/p')&id=1jwRgRHPJrKABOk7wImKONTtUupV7yJ9b\" -O data_bulk.tar.gz\n",
    "rm -rf /tmp/cookies.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03a7cb13-1651-4ba6-b9df-ff0a47d059f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "!tar -xzf data_bulk.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc1052a8-1a16-490d-92dd-1e2ccc3ec366",
   "metadata": {},
   "source": [
    "## 2. Prepare the input data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f734203a-e340-443d-89e6-1068f0bda8cf",
   "metadata": {},
   "source": [
    "The input data is the feature matrix from 10x sc-multiome data and Cell annotation/cell type label which includes: \n",
    "- Single-cell multiome data including matrix.mtx.gz, features.tsv.gz, and barcodes.tsv.gz.\n",
    "- Cell annotation/cell type label if you need the cell type-specific gene regulatory network (PBMC_label.txt in our example)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e5c9c60-71f1-4615-9132-af39747a6801",
   "metadata": {},
   "source": [
    "LINGER is designed for single-cell multiome data, this means that for each cell we have both :\n",
    "- gene expression (scRNA-seq) -> \"Which genes are expressed in each cell ?\"\n",
    "- chromatin accessibility (scATAC-seq) -> \"Which regulatory DNA regions are accessible in the same cell ?\"\n",
    "\n",
    "We use the public 10x Genomics PBMC multiome dataset \"*pbmc_granulocyte_sorted_10k*\"\n",
    "(≈10,000 peripheral blood mononuclear cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0748bf09-c8f8-412c-8ef0-5539db181785",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Download PBMC multiome data\n",
    "mkdir -p data\n",
    "wget -O data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.tar.gz \\\n",
    "https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.tar.gz\n",
    "\n",
    "tar -xzvf data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.tar.gz\n",
    "mv filtered_feature_bc_matrix data/\n",
    "gzip -d data/filtered_feature_bc_matrix/*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "112a5721-6e64-45da-aed5-62f39a078084",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Download cell type annotation labels\n",
    "wget --load-cookies /tmp/cookies.txt \"https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\\1\\n/p')&id=17PXkQJr8fk0h90dCkTi3RGPmFNtDqHO_\" -O PBMC_label.txt\n",
    "mv PBMC_label.txt data/\n",
    "rm -rf /tmp/cookies.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4898c853-e8c3-47ed-990e-77e691e94b18",
   "metadata": {},
   "source": [
    "We end up with the following files :\n",
    "\n",
    "- *matrix.mtx* : A sparse count matrix with both RNA and ATAC counts mixed together (180488 features x 11898 cells)\n",
    "    1. Rows = features (genes + peaks)\n",
    "    2. Columns = cells\n",
    "    3. Values = number of reads <br><br>\n",
    "\n",
    "- *features.tsv* : Each row describes what a row in *matrix.mtx* means, it splits the matrix into RNA and ATAC parts (180488 rows)\n",
    "\n",
    "- *barcodes.tsv* : One barcode per column in *matrix.mtx*, each barcode = one signle cell (11898 barcodes)\n",
    "\n",
    "- *PBMC_label.txt* : cell annotation metadata, tells us which cell type each barcode corresponds to\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb328403-ca18-411e-8983-d267a0f603ae",
   "metadata": {},
   "source": [
    "## 3. Install LINGIER + conda env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f1aa2f0-9288-4211-84b9-623b79032c89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell 4: Install LINGER and dependencies (run in notebook terminal or a separate conda env)\n",
    "# Uncomment and run once\n",
    "\"\"\"\n",
    "!conda create -n LINGER python=3.10 -y\n",
    "!conda activate LINGER\n",
    "!pip install LingerGRN==1.105\n",
    "!conda install -c bioconda bedtools conda-forge scanpy pandas scipy -y\n",
    "\n",
    "!conda install -c conda-forge scanpy=1.9 pandas scipy matplotlib=3.7 anndata=0.9 -y\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a16a4f6-e089-4238-9282-ecc667ce51ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda info --envs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2329ec8b-e2ba-4e02-8450-820335843f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda list -n LINGER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9965f21b-8387-4b35-8ffb-5ecf9b8b66f7",
   "metadata": {},
   "source": [
    "## 4. About the AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad680470-dc51-4c9d-8055-c54430b1d1d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preprocess - load sc-multiome data into AnnData\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import scipy.io\n",
    "from LingerGRN.preprocess import *\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "matrix = scipy.io.mmread('data/filtered_feature_bc_matrix/matrix.mtx')\n",
    "features = pd.read_csv('data/filtered_feature_bc_matrix/features.tsv', sep='\\t', header=None)\n",
    "barcodes = pd.read_csv('data/filtered_feature_bc_matrix/barcodes.tsv', sep='\\t', header=None)\n",
    "label = pd.read_csv('data/PBMC_label.txt', sep='\\t', header=0)\n",
    "\n",
    "adata_RNA, adata_ATAC = get_adata(matrix, features, barcodes, label)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5ad5b5ca-2ef0-4c41-969b-daf2253411ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "anndata._core.anndata.AnnData"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(adata_ATAC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78b110fd-df98-46e2-babb-e5b5c56ce8e0",
   "metadata": {},
   "source": [
    "An [AnnData](https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData) object (annotated data) *adata* stores a data matrix adata.X, annotation of observations adata.obs and variables adata.var as pd.DataFrame and unstructured annotation adata.uns as dict. Names of observations and variables can be accessed via adata.obs_names and adata.var_names, respectively.\n",
    "\n",
    "adata_RNA & adata_ATAC are two aligned AnnData objects of same cells with different features (genes -> RNA and peaks -> ATAC)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63236af7-aa19-40f5-84f1-409c128861e6",
   "metadata": {},
   "source": [
    "### A. adata_RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8c103b4e-c479-4d09-824a-485f5d170213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 9543 × 36601\n",
       "    obs: 'barcode', 'sample', 'label', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt'\n",
       "    var: 'gene_ids', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09d4c51b-a20e-4c75-b8fb-f64ca3d12dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_RNA.X                   # cells x genes, contains RNA counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "61a710ad-4073-41d7-a91c-81dfa6e11849",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample</th>\n",
       "      <th>label</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11249</th>\n",
       "      <td>TTAGGCTAGCGAGTAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>classical monocytes</td>\n",
       "      <td>2086</td>\n",
       "      <td>4537.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6572</th>\n",
       "      <td>GCAAGTGCATAAGCAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD8 T cells</td>\n",
       "      <td>1672</td>\n",
       "      <td>3312.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>AATTGACGTGCTAGAC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD8 T cells</td>\n",
       "      <td>1689</td>\n",
       "      <td>3454.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>AATAGAGGTGCTCACC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>myeloid DC</td>\n",
       "      <td>3311</td>\n",
       "      <td>9175.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11567</th>\n",
       "      <td>TTGGCTTGTGAGCGAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD8 T cells</td>\n",
       "      <td>1436</td>\n",
       "      <td>2714.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  barcode  sample                label  n_genes_by_counts  \\\n",
       "11249  TTAGGCTAGCGAGTAA-1       1  classical monocytes               2086   \n",
       "6572   GCAAGTGCATAAGCAA-1       1    naive CD8 T cells               1672   \n",
       "663    AATTGACGTGCTAGAC-1       1    naive CD8 T cells               1689   \n",
       "523    AATAGAGGTGCTCACC-1       1           myeloid DC               3311   \n",
       "11567  TTGGCTTGTGAGCGAA-1       1    naive CD8 T cells               1436   \n",
       "\n",
       "       total_counts  total_counts_mt  pct_counts_mt  \n",
       "11249        4537.0              0.0            0.0  \n",
       "6572         3312.0              0.0            0.0  \n",
       "663          3454.0              0.0            0.0  \n",
       "523          9175.0              0.0            0.0  \n",
       "11567        2714.0              0.0            0.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.obs.head()          # cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "06e7c9be-d7ef-4803-955a-732a37651fd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>False</td>\n",
       "      <td>61</td>\n",
       "      <td>0.006916</td>\n",
       "      <td>99.360788</td>\n",
       "      <td>66.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ids     mt  n_cells_by_counts  mean_counts  \\\n",
       "MIR1302-2HG  MIR1302-2HG  False                  0     0.000000   \n",
       "FAM138A          FAM138A  False                  0     0.000000   \n",
       "OR4F5              OR4F5  False                  0     0.000000   \n",
       "AL627309.1    AL627309.1  False                 61     0.006916   \n",
       "AL627309.3    AL627309.3  False                  0     0.000000   \n",
       "\n",
       "             pct_dropout_by_counts  total_counts  \n",
       "MIR1302-2HG             100.000000           0.0  \n",
       "FAM138A                 100.000000           0.0  \n",
       "OR4F5                   100.000000           0.0  \n",
       "AL627309.1               99.360788          66.0  \n",
       "AL627309.3              100.000000           0.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.var.head()          # genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "56307799-d4bf-44ad-972b-e88842b63c13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9543, 36601)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.X.shape             # 9543 cells x 36601 genes  (also non protein coding genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4e6e5387-71bf-4880-b893-ddfdf422379b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18510894"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.X.nnz               # nb of non zero entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7ff4a1a3-61be-4e72-9aac-7817899071ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0529967843327702"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# those non zero entries represent a very small fraction of the matrix\n",
    "adata_RNA.X.nnz / (adata_RNA.shape[0] * adata_RNA.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "50f01fd4-7d06-459a-af7f-c6e7e2d17ab9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.,\n",
       "        2., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        2., 0., 0., 0., 0., 0., 6., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0.]], dtype=float32)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.X[7, :100].toarray() # 7th cell, 100 first genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4a81ef2a-f5db-45e4-8cd2-917bf5b6625b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "barcode               CGAAGCGAGGGCTAAA-1\n",
       "sample                                 1\n",
       "label                classical monocytes\n",
       "n_genes_by_counts                   2131\n",
       "total_counts                      4341.0\n",
       "total_counts_mt                      0.0\n",
       "pct_counts_mt                        0.0\n",
       "Name: 4300, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.obs.iloc[7]          # metadata of the 7th obs (cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "18aa3955-7867-4114-a7ce-96cfbd7a4b13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_ids                 MIR1302-2HG\n",
       "mt                             False\n",
       "n_cells_by_counts                  0\n",
       "mean_counts                      0.0\n",
       "pct_dropout_by_counts          100.0\n",
       "total_counts                     0.0\n",
       "Name: MIR1302-2HG, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.var.iloc[0]          # metadata of gene 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "534a406c-7035-4838-9571-56c6a9934f9d",
   "metadata": {},
   "source": [
    "### B. adata_ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e5ebc5d4-a69b-454a-9379-6314d8effef1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 9543 × 143887\n",
       "    obs: 'barcode', 'sample', 'label'\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "cb83c579-1344-4651-8af0-58e445260092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11249</th>\n",
       "      <td>TTAGGCTAGCGAGTAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>classical monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6572</th>\n",
       "      <td>GCAAGTGCATAAGCAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD8 T cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>AATTGACGTGCTAGAC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD8 T cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>AATAGAGGTGCTCACC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>myeloid DC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11567</th>\n",
       "      <td>TTGGCTTGTGAGCGAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>naive CD8 T cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  barcode  sample                label\n",
       "11249  TTAGGCTAGCGAGTAA-1       1  classical monocytes\n",
       "6572   GCAAGTGCATAAGCAA-1       1    naive CD8 T cells\n",
       "663    AATTGACGTGCTAGAC-1       1    naive CD8 T cells\n",
       "523    AATAGAGGTGCTCACC-1       1           myeloid DC\n",
       "11567  TTGGCTTGTGAGCGAA-1       1    naive CD8 T cells"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC.obs.head()         # cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "82d46926-98b2-4330-9a28-5431f4aab088",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36601</th>\n",
       "      <td>chr1:9790-10675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36602</th>\n",
       "      <td>chr1:180599-181702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36603</th>\n",
       "      <td>chr1:191168-192093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36604</th>\n",
       "      <td>chr1:267565-268455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36605</th>\n",
       "      <td>chr1:270876-271770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_ids\n",
       "36601     chr1:9790-10675\n",
       "36602  chr1:180599-181702\n",
       "36603  chr1:191168-192093\n",
       "36604  chr1:267565-268455\n",
       "36605  chr1:270876-271770"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC.var.head()          # peak regions (= genomic region of open chromatin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "9bccec5b-ad3d-4a02-ad73-168b8ad39081",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9543, 143887)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC.shape              # cells x peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ca43f3a7-8230-4858-bc92-d45a11fdb8b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 2., 0., 0., 0., 0.,\n",
       "        2., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 2.,\n",
       "        0., 0., 0., 0., 0., 2., 0., 2., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 2., 0., 0., 2., 0., 0., 4., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "        0., 0., 6., 0., 0., 0., 0., 0., 2., 0., 0., 0., 2., 0., 1., 4.,\n",
       "        0., 0., 4., 0.]], dtype=float32)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC.X[1, :100].toarray()     # cell 1, 100 first peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5db7919d-ae5d-46e0-b2ea-5a1bc72b4b23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "barcode        CAGCTATAGTTGGCCA-1\n",
       "sample                          1\n",
       "label      intermediate monocytes\n",
       "Name: 3291, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC.obs.iloc[1]             # cell 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "310c6e1f-7cdc-4605-a56a-c5884ba80190",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_ids    chr1:778283-779200\n",
       "Name: 36612, dtype: object"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ATAC.var.iloc[11]            # peak 11"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "809f247d-4448-47e4-8817-85ba23021a96",
   "metadata": {},
   "source": [
    "From above, cell 1 has a peak in chr1:778283-779200 (since there is a 2 at *adata_ATAC.X[1,11]*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e031dd6d-510d-441a-b1b7-7683eb89d401",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata_RNA.X[i, j]             # cell i, gene j : Gene expression count of gene j in cell i\n",
    "#adata_ATAC.X[i, k]            # cell i, peak k : Chromatin accessibility count of peak k in cell i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "0b58e18b-e1bf-437b-be98-4aed1f5b672c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA.X[7, 42]             # gives 1 : cell 7 has 1 RNA molecule for gene 42"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d854fe0-e687-4516-add5-87d6f3ca20ba",
   "metadata": {},
   "source": [
    "## 5. Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf3f0d9-8d0a-4dba-ae8f-eb8433cf2daf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter low-count cells and genes\n",
    "\n",
    "# Keep only cells that have ≥ 200 detected genes\n",
    "sc.pp.filter_cells(adata_RNA, min_genes=200)\n",
    "\n",
    "# Keep only genes expressed in ≥ 3 cells\n",
    "sc.pp.filter_genes(adata_RNA, min_cells=3)\n",
    "\n",
    "sc.pp.filter_cells(adata_ATAC, min_genes=200)\n",
    "sc.pp.filter_genes(adata_ATAC, min_cells=3)\n",
    "\n",
    "# Keep only cells present in both RNA and ATAC\n",
    "selected_barcode = list(set(adata_RNA.obs['barcode'].values) & set(adata_ATAC.obs['barcode'].values))\n",
    "\n",
    "barcode_idx = pd.DataFrame(range(adata_RNA.shape[0]), index=adata_RNA.obs['barcode'].values)\n",
    "adata_RNA = adata_RNA[barcode_idx.loc[selected_barcode][0]]\n",
    "\n",
    "barcode_idx = pd.DataFrame(range(adata_ATAC.shape[0]), index=adata_ATAC.obs['barcode'].values)\n",
    "adata_ATAC = adata_ATAC[barcode_idx.loc[selected_barcode][0]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a777f00-fdb5-4b9b-869d-57b5f98b3dae",
   "metadata": {},
   "source": [
    "### About pseudo-bulking\n",
    "\n",
    "Pseudo-bulk means \"Combine many single cells into a “fake bulk sample” by summing or averaging their counts\"\n",
    "Since single cell is noisy and extremely sparse, it is better to work with aggregated signals across groups of cells (=metacells).  \n",
    "- ``singlepseudobulk = true`` : Collapse all cells in this sample into ONE pseudobulk profile. This gives following dimensions : \n",
    "    - TG_pseudobulk_temp : (n_genes × 1)\n",
    "    - RE_pseudobulk_temp : (n_peaks × 1) <br><br>\n",
    "      \n",
    "- ``singlepseudobulk = false`` : First cluster cells → then make multiple pseudobulks (metacells), used when we don't have many samples. This will create *K* clusters of cells, or *K* metacells\n",
    "    - TG_pseudobulk_temp : (n_genes × k_metacells)\n",
    "    - RE_pseudobulk_temp : (n_peaks × k_metacells) <br><br>\n",
    "      \n",
    "- Why is this needed ? GRN inference needs many samples (columns).\n",
    "    - If you already have many samples → 1 bulk per sample is enough\n",
    "    - If you have few samples → create metacells to increase sample count\n",
    "      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb513ef2-844e-4187-a65d-045b4b79398b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate pseudo-bulk/metacell\n",
    "import os\n",
    "from LingerGRN.pseudo_bulk import *\n",
    "\n",
    "samplelist=list(set(adata_ATAC.obs['sample'].values)) # sample is generated from cell barcode \n",
    "tempsample=samplelist[0]\n",
    "\n",
    "TG_pseudobulk=pd.DataFrame([])\n",
    "RE_pseudobulk=pd.DataFrame([])\n",
    "\n",
    "n_samples = adata_RNA.obs['sample'].nunique()\n",
    "singlepseudobulk = (n_samples > 10)\n",
    "#singlepseudobulk = (adata_RNA.obs['sample'].unique().shape[0]*adata_RNA.obs['sample'].unique().shape[0]>100)\n",
    "\n",
    "# here samplelist = [1], singlepseudobulk = False (there is only one sample)\n",
    "for tempsample in samplelist:\n",
    "\n",
    "    # get cells from only tempsample\n",
    "    adata_RNAtemp = adata_RNA[adata_RNA.obs['sample' ] == tempsample]\n",
    "    adata_ATACtemp = adata_ATAC[adata_ATAC.obs['sample'] == tempsample]\n",
    "\n",
    "    TG_pseudobulk_temp, RE_pseudobulk_temp = pseudo_bulk(adata_RNAtemp, adata_ATACtemp, singlepseudobulk)  \n",
    "    \n",
    "    TG_pseudobulk = pd.concat([TG_pseudobulk, TG_pseudobulk_temp], axis=1)\n",
    "    RE_pseudobulk = pd.concat([RE_pseudobulk, RE_pseudobulk_temp], axis=1)\n",
    "    \n",
    "    RE_pseudobulk[RE_pseudobulk > 100] = 100\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f0cfe1e-8f21-4eab-9aa2-34077daf6562",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('data/'):\n",
    "    os.mkdir('data/')\n",
    "    \n",
    "adata_ATAC.write('data/adata_ATAC.h5ad')\n",
    "adata_RNA.write('data/adata_RNA.h5ad')\n",
    "\n",
    "TG_pseudobulk=TG_pseudobulk.fillna(0)\n",
    "RE_pseudobulk=RE_pseudobulk.fillna(0)\n",
    "\n",
    "pd.DataFrame(adata_ATAC.var['gene_ids']).to_csv('data/Peaks.txt',header=None,index=None)\n",
    "\n",
    "TG_pseudobulk.to_csv('data/TG_pseudobulk.tsv')\n",
    "RE_pseudobulk.to_csv('data/RE_pseudobulk.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e1ecb67-a0f9-4d67-85aa-1137647edb91",
   "metadata": {},
   "source": [
    "## 6. Training the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da174406-b2c7-4838-968c-6b00263e23f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/emile/Documents/UCL/Memoire/Lingier/LINGER_data/data_bulk/'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "from LingerGRN.preprocess import *\n",
    "\n",
    "Datadir = os.path.join(os.getcwd(), 'LINGER_data/')\n",
    "GRNdir = Datadir + 'data_bulk/'\n",
    "genome = 'hg38'\n",
    "outdir = '/LINGER_output/'  # output directory\n",
    "\n",
    "preprocess(TG_pseudobulk, RE_pseudobulk, GRNdir, genome, method, outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49aa6e06-5199-4db0-8ebb-a42143e31041",
   "metadata": {},
   "outputs": [],
   "source": [
    "import LingerGRN.LINGER_tr as LINGER_tr\n",
    "\n",
    "method = 'baseline'    # or 'LINGER'\n",
    "activef='ReLU' # active function chose from 'ReLU','sigmoid','tanh'\n",
    "LINGER_tr.training(GRNdir,method,outdir,activef,'Human')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7928d8b-8937-4b55-ae40-9d1570089d17",
   "metadata": {},
   "source": [
    "## 7. Cell population gene regulatory network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5df7c578-3725-45e8-8109-5be5c652a518",
   "metadata": {},
   "source": [
    "### 7.1 TF binding potential (TF-RE)\n",
    "The output is 'cell_population_TF_RE_binding.txt', a matrix of the TF-RE binding score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d39a68e4-b60a-4720-a73b-062d014c7cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import LingerGRN.LL_net as LL_net\n",
    "LL_net.TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab01b819-658e-4833-be5d-4a08d921cc49",
   "metadata": {},
   "source": [
    "### 7.2 cis-regulatory network (RE-TG)\n",
    "The output is 'cell_population_cis_regulatory.txt' with 3 columns: region, target gene, cis-regulatory score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a412ca2-5ba5-4130-80e2-bd14486f82ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc94a2a-74a1-4062-bf10-26c6d14caae8",
   "metadata": {},
   "source": [
    "### 7.3 trans-regulatory network (TF-TG)\n",
    "The output is 'cell_population_trans_regulatory.txt', a matrix of the trans-regulatory score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52206b73-71b7-44bb-9ee6-999e79528bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.trans_reg(GRNdir,method,outdir,genome)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (LINGER)",
   "language": "python",
   "name": "linger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
